gcs/constraints/smart_table.cc

```diff
@@ -0,0 +1,642 @@
+                if (visited) continue;
+                vector<vector<SmartEntry>> entry_tree;
+                entry_tree.emplace_back();
+                // Recursively build the tree starting from this node
+                build_tree(var, 0, entry_tree, node_visited, adjacent_edges);
+                forest.emplace_back(entry_tree);
+            }
+
+            forests.emplace_back(forest);
+        }
+        return forests;
+    }
+}
+
+// For PB model
+auto make_binary_entry_flag(State & state, Propagators & propagators, const IntegerVariableID & var_1, const IntegerVariableID & var_2, const ConstraintType & c) -> ProofFlag
+{
+    ProofFlag flag, flag_lt, flag_gt; // Is this initialisation okay?
+    WeightedPseudoBooleanTerms var_1_minus_var_2 = {{1_i, var_1}, {-1_i, var_2}};
+    WeightedPseudoBooleanTerms var_2_minus_var_1 = {{1_i, var_2}, {-1_i, var_1}};
+    switch (c) {
+    case EQUAL:
+        // f => var1 == var2
+        flag = propagators.create_proof_flag("bin_eq");
+        propagators.define_pseudoboolean_eq(state, move(var_1_minus_var_2), 0_i, flag);
+
+        // !f => var1 != var2
+        flag_lt = propagators.create_proof_flag("lt");
+        flag_gt = propagators.create_proof_flag("gt");
+        propagators.define_pseudoboolean_ge(state, move(var_1_minus_var_2), 1_i, flag_gt);
+        propagators.define_pseudoboolean_ge(state, move(var_2_minus_var_1), 0_i, ! flag_gt);
+        propagators.define_pseudoboolean_ge(state, move(var_2_minus_var_1), 1_i, flag_lt);
+        propagators.define_pseudoboolean_ge(state, move(var_1_minus_var_2), 0_i, ! flag_lt);
+        propagators.define_pseudoboolean_ge(state, {{1_i, flag_lt}, {1_i, flag_gt}}, 1_i, ! flag);
+        return flag;
+    case GREATER_THAN:
+        flag = propagators.create_proof_flag("bin_gt");
+        propagators.define_pseudoboolean_ge(state, move(var_1_minus_var_2), 1_i, flag);
+        propagators.define_pseudoboolean_ge(state, move(var_2_minus_var_1), 0_i, ! flag);
+        return flag;
+    case LESS_THAN:
+        flag = propagators.create_proof_flag("bin_lt");
+        propagators.define_pseudoboolean_ge(state, move(var_2_minus_var_1), 1_i, flag);
+        propagators.define_pseudoboolean_ge(state, move(var_1_minus_var_2), 0_i, ! flag);
+        return flag;
+    case LESS_THAN_EQUAL:
+        propagators.define_pseudoboolean_ge(state, move(var_2_minus_var_1), 0_i, flag);
+        propagators.define_pseudoboolean_ge(state, move(var_1_minus_var_2), 1_i, ! flag);
+        return flag;
+    case NOT_EQUAL:
+        // !f => var1 == var2
+        flag = propagators.create_proof_flag("bin_eq");
+        propagators.define_pseudoboolean_eq(state, move(var_1_minus_var_2), 0_i, ! flag);
+
+        // f => var1 != var2
+        flag_lt = propagators.create_proof_flag("lt");
+        flag_gt = propagators.create_proof_flag("gt");
+        propagators.define_pseudoboolean_ge(state, move(var_1_minus_var_2), 1_i, flag_gt);
```

Comment:
Again, I think you're reusing moved variables.

Reply:
Fixed in c62e8914eb9e01df2a147f01713ff505d7e905e0